crates/runtime/src/libcalls.rs

```diff
@@ -507,7 +508,11 @@ unsafe fn validate_atomic_addr(
     memory: MemoryIndex,
     addr: usize,
 ) -> Result<(), TrapCode> {
-    if addr > instance.get_memory(memory).current_length() {
+    let mem = instance.get_memory(memory);
+    if addr < mem.base as usize {
+        return Err(TrapCode::HeapOutOfBounds);
+    }
+    if addr >= mem.base.add(mem.current_length.load(Ordering::Relaxed)) as *const _ as usize {
```

Comment:
Does this also need to account for access sizes larger than 1? For example a 8 byte access where the address is the last byte of the memory should not be allowed.

Reply:
Misaligned access will trap. This check is done in JIT code: https://github.com/bytecodealliance/wasmtime/blob/32a7593c949a17f20c2a718e1cbf0d8c449b0c47/cranelift/wasm/src/code_translator.rs#L1070-L1071

Reply:
That would necessitate all users of this function to trap on misalignment. Currently all callers of this function seem to immediately be followed by an unimplemented trap. Once they are implemented it is easy to miss this requirement as misaligned atomic accesses in rust are not guaranteed to panic.